app/cluster/dynamic.go

```diff
@@ -0,0 +1,104 @@
+package cluster
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/rudderlabs/rudder-server/utils/types/servermode"
+)
+
+type modeProvider interface {
+	ServerMode() <-chan servermode.ModeAck
+}
+
+type Lifecycle interface {
+	Start()
+	Stop()
+}
+
+type Dynamic struct {
+	Provider modeProvider
+
+	GatewayDB     Lifecycle
+	RouterDB      Lifecycle
+	BatchRouterDB Lifecycle
+	ErrorDB       Lifecycle
+
+	Processor Lifecycle
+	Router    Lifecycle
+
+	currentMode servermode.Mode
+}
+
+func (d *Dynamic) Run(ctx context.Context) error {
+	for {
+		select {
+		case <-ctx.Done():
+			if d.currentMode == servermode.NormalMode {
+				d.stop()
+			}
+			return nil
+		case newMode := <-d.Provider.ServerMode():
+			err := d.handleModeChange(newMode.Mode())
+			if err != nil {
+				return err
+			}
+			newMode.Ack()
+		}
+	}
+}
+
+func (d *Dynamic) start() {
```

Comment:
Is there any specific reasoning behind the current start/stop order?
Normally, one would expect the stop order to be the inverse of the start order.

A start strategy that just came into mind is to start based on the inverse order of the event processing steps, e.g.

1. RouterDB
2. Router
4. BatchRouterDB
5. BatchRouter
6. ErrorDB
7. GatewayDB
8. Processor

In the above list I would also expect to see a Gateway too. What would happen if we had a running Gateway operating against a stopped GatewayDB?

Reply:
> Normally, one would expect the stop order to be the inverse of the start order.

That is the idea, well in our case as long as dbs start before the processor/router we are set. We are thinking of start/stop them concurrently to speed up things.

Also, router includes batchrouter.

Reply:
> In the above list I would also expect to see a Gateway too. What would happen if we had a running Gateway operating against a stopped GatewayDB?

This part is a bit more complicated, the short answer is: For now gateway is going to run in another process. Dynamic server mode in embedded setup, needs more effort and we are going to tackle it later.

Reply:
won't affect because processor or router will only be started after DBs